internal/manifest/oci_index_test.go

```diff
@@ -45,3 +49,113 @@ func TestOCI1IndexFromManifest(t *testing.T) {
+			},
+		},
+		{
+			listFile: "oci1.index.zstd-selection2.json",
+			// out of list where first instance is gzip , select the first occurance of zstd out of many
+			matchedInstances: []expectedMatch{
+				{"amd64", "", "sha256:ac9bbf71dacfad2bfbe48f7d46ca279b1b3b351a6c814db760030015167405b8"},
+			},
+			prefferedGzipInstances: []expectedMatch{
+				{"amd64", "", "sha256:88cb27f194cce3b06994b4af2dd8a5642f4f294b71bda5b74a3fbb74a9522510"},
+			},
+			unmatchedInstances: []string{
+				"unmatched",
+			},
+		},
+		{ // Focus on ARM variant field testing
+			listFile: "schema2list-variants.json",
+			matchedInstances: []expectedMatch{
+				{"amd64", "", "sha256:59eec8837a4d942cc19a52b8c09ea75121acc38114a2c68b98983ce9356b8610"},
+				{"arm", "v7", "sha256:f365626a556e58189fc21d099fc64603db0f440bff07f77c740989515c544a39"},
+				{"arm", "v6", "sha256:f365626a556e58189fc21d099fc64603db0f440bff07f77c740989515c544a39"},
+				{"arm", "v5", "sha256:c84b0a3a07b628bc4d62e5047d0f8dff80f7c00979e1e28a821a033ecda8fe53"},
+				{"arm", "", "sha256:c84b0a3a07b628bc4d62e5047d0f8dff80f7c00979e1e28a821a033ecda8fe53"},
+				{"arm", "unrecognized-present", "sha256:bcf9771c0b505e68c65440474179592ffdfa98790eb54ffbf129969c5e429990"},
+				{"arm", "unrecognized-not-present", "sha256:c84b0a3a07b628bc4d62e5047d0f8dff80f7c00979e1e28a821a033ecda8fe53"},
+			},
+			unmatchedInstances: []string{
+				"unmatched",
+			},
+		},
+		{
+			listFile: "oci1index.json",
+			matchedInstances: []expectedMatch{
+				{"amd64", "", "sha256:5b0bcabd1ed22e9fb1310cf6c2dec7cdef19f0ad69efa1f392e94a4333501270"},
+				{"ppc64le", "", "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f"},
+			},
+			unmatchedInstances: []string{
+				"unmatched",
+			},
+		},
+	} {
+		rawManifest, err := os.ReadFile(filepath.Join("..", "..", "internal", "manifest", "testdata", manifestList.listFile))
+		require.NoError(t, err)
+		list, err := ListFromBlob(rawManifest, GuessMIMEType(rawManifest))
+		require.NoError(t, err)
+		// Match found and Gzip preference false
+		for _, match := range manifestList.matchedInstances {
+			testName := fmt.Sprintf("%s %q+%q", manifestList.listFile, match.arch, match.variant)
+			digest, err := list.ChooseInstanceByCompression(&types.SystemContext{
+				ArchitectureChoice: match.arch,
+				VariantChoice:      match.variant,
+				OSChoice:           "linux",
+			}, types.NewOptionalBool(false))
+			require.NoError(t, err, testName)
+			assert.Equal(t, match.instanceDigest, digest, testName)
+		}
+		// Not found and Gzip preference false
+		for _, arch := range manifestList.unmatchedInstances {
+			_, err := list.ChooseInstanceByCompression(&types.SystemContext{
+				ArchitectureChoice: arch,
+				OSChoice:           "linux",
+			}, types.NewOptionalBool(false))
+			assert.Error(t, err)
+		}
+		// Match found and Gzip preference true
+		for _, match := range manifestList.prefferedGzipInstances {
```

Comment:
Make the `preferGzip` input a new field in `expectedMatch` (https://github.com/containers/image/pull/1789#discussion_r1101831378); then this loop, and `preferredGzipInstances`, don’t need to exist, _and_ we can also, possibly, test `OptionalBoolUndefined` more easily.